src/rdkafka_broker.c

```diff
@@ -4206,73 +4208,115 @@ static int rd_kafka_broker_thread_main (void *arg) {
                                 rd_kafka_broker_serve(rkb, 1000);
 
                         /* Throttle & jitter reconnects to avoid
-                         * thundering horde of reconnecting clients after
-                         * a broker / network outage. Issue #403 */
+                        * thundering horde of reconnecting clients after
+                        * a broker / network outage. Issue #403 */
                         backoff = rd_kafka_broker_reconnect_backoff(rkb,
-                                                                    rd_clock());
+                                                                rd_clock());
                         if (backoff > 0) {
                                 rd_rkb_dbg(rkb, BROKER, "RECONNECT",
-                                           "Delaying next reconnect by %dms",
-                                           backoff);
+                                        "Delaying next reconnect by %dms",
+                                        backoff);
                                 rd_kafka_broker_serve(rkb, (int)backoff);
                                 continue;
                         }
 
-			/* Initiate asynchronous connection attempt.
-			 * Only the host lookup is blocking here. */
-			if (rd_kafka_broker_connect(rkb) == -1) {
-				/* Immediate failure, most likely host
-				 * resolving failed.
-				 * Try the next resolve result until we've
-				 * tried them all, in which case we sleep a
-				 * short while to avoid busy looping. */
-				if (!rkb->rkb_rsal ||
-                                    rkb->rkb_rsal->rsal_cnt == 0 ||
-                                    rkb->rkb_rsal->rsal_curr + 1 ==
-                                    rkb->rkb_rsal->rsal_cnt)
+                        /*
+                        * SASL/OAUTHBEARER is unable to connect unless a valid token is
+                        * available, and a valid token CANNOT be available unless/until
+                        * an initial token retrieval succeeds, so wait for this
+                        * precondition if necessary.
+                        */
+                        if (rkb->rkb_rk->rk_oauthbearer) {
+                                int has_token = 0;
+                                int est_total_wait_ms = 0;
+                                int max_wait_ms = 30000;
+                                while (!has_token && est_total_wait_ms < max_wait_ms) {
+                                        rwlock_rdlock(&rkb->rkb_rk->rk_oauthbearer->refresh_lock);
+                                        has_token = rkb->rkb_rk->rk_oauthbearer->token_value != NULL;
+                                        rwlock_rdunlock(&rkb->rkb_rk->rk_oauthbearer->refresh_lock);
+                                        if (has_token) {
+                                                rd_rkb_dbg(rkb, BROKER, "BRKMAIN",
+                                                        "Main broker thread has a token");
```

Comment:
"Main broker thread" is not useful for the user. Better to say "OAUTHBEARER token available"

Reply:
👍 Fixed